framework/src/bundle/BundleRegistry.cpp

```diff
@@ -179,140 +269,201 @@ std::vector<Bundle> BundleRegistry::Install(const std::string& location,
       }
       
       // Re-acquire the range because while this thread was waiting, the installing
       // thread made a modification to bundles.v
+      l.Lock();
       bundlesAtLocationRange = (bundles.Lock(), bundles.v.equal_range(location));
       l.UnLock();
 
       std::vector<Bundle> resultingBundles;
-      std::vector<std::shared_ptr<BundlePrivate>> alreadyInstalled;
-      GetAlreadyInstalledBundlesAtLocation(bundlesAtLocationRange, resultingBundles, alreadyInstalled);
+      std::vector<std::string> alreadyInstalled;
+      auto resCont = GetAlreadyInstalledBundlesAtLocation(bundlesAtLocationRange
+                                                          , location
+                                                          , bundleManifest
+                                                          , resultingBundles
+                                                          , alreadyInstalled);
 
       std::vector<Bundle> newBundles;
-      
       {
         // create instance of clean-up object to ensure RAII
-        InitialBundleMapCleanup cleanup([this, &l, &location](){
-          DecrementInitialBundleMapRef(l, location);
-        });
+        scope_guard cleanup = [this, &l, &location]()
+                              {
+                                DecrementInitialBundleMapRef(l, location);
+                              };
 
         // Perform the install
-        newBundles = Install0(location, alreadyInstalled, caller);
+        newBundles = Install0(location, resCont, alreadyInstalled, bundleManifest);
       }
 
       resultingBundles.insert(resultingBundles.end(), newBundles.begin(), newBundles.end());
-      if (resultingBundles.empty()) {
+      if (resultingBundles.empty())
+      {
         throw std::runtime_error("All bundles rejected by a bundle hook");
-      } else {
-        return resultingBundles;
       }
+      return resultingBundles;
     }
   }
 }
 
-std::vector<Bundle> BundleRegistry::Install0(
-  const std::string& location,
-  const std::vector<std::shared_ptr<BundlePrivate>>& exclude,
-  BundlePrivate* /*caller*/)
+std::vector<Bundle> BundleRegistry::Install0(const std::string& location
+                                             , const std::shared_ptr<BundleResourceContainer>& resCont
+                                             , const std::vector<std::string>& alreadyInstalled
+                                             , const ManifestT& bundleManifest)
 {
-  std::vector<Bundle> res;
+  namespace cppms = cppmicroservices;
+  using cppms::AnyMap;
+  using cppms::any_cast;
+  using cppms::any_map;
+  
+  std::vector<Bundle> installedBundles;
   std::vector<std::shared_ptr<BundleArchive>> barchives;
+  scope_guard purge_on_error = [&barchives]()
+                               {
+                                 for (auto& ba : barchives) {
+                                   ba->Purge();
+                                 }
+                               };
```

Comment:
i don't think this `scope_guard` is necessary since all the code which could fail is already contained in a `try/catch` and the cleanup action was already done in the `catch` block.

Reply:
Not sure I agree with this one. I like the idea of having the purging code right next to the variable that's being cleaned up (referenced in the captures). Also, I think this protects against future exception handling code from accidentally forgetting to purge. However, if you're adamant about this removal, I'll take it out.

Reply:
I understand. in the same vein, I'm not a fan of having `dismiss` called so far away from the instantiation of the `scope_guard`.   In my opinion this change is unnecessary. If this code was re-factored to remove the try/catch I would see the merit in using `scope_guard`.

Reply:
OK. Done.

Reply:
Done.